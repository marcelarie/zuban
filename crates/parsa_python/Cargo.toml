[package]
name = "parsa_python"
version.workspace = true
edition.workspace = true
rust-version.workspace = true
publish = false
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
test_utils.workspace = true
