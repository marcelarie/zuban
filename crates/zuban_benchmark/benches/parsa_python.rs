use criterion::{criterion_group, criterion_main, Criterion};

use parsa_python_cst::Tree;

fn parse_file(copies: usize) {
    let some_code = utils::dedent(
        r#"
        import os
        def x(a, b, c):
            try:
                return a
            except Exception:
                return [1, 2, 3, 4, 5, 6]
        class C:
            x: int
            y = 1 # type: int
            z: dict[int, list[str]] = {1: ["", r''], 2: [f""" {1} """]}
            def f(self, x: list[str], y: int): ...
    "#,
    );
    assert_eq!(some_code.len(), 262);
    Tree::parse(some_code.repeat(copies).into());
}

fn bench_parser(c: &mut Criterion) {
    c.bench_function("parse big file", |b| b.iter(|| parse_file(1000)));
    c.bench_function("parse small file", |b| b.iter(|| parse_file(1)));
}

// Register the benchmarks
criterion_group!(benches, bench_parser);
criterion_main!(benches);
