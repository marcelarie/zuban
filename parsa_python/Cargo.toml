[package]
name = "parsa_python"
version = "0.1.0"
edition.workspace = true
rust-version.workspace = true
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
