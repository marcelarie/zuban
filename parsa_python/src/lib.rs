use parsa::{create_parser,create_node,create_token};


create_token!(PythonToken, enum TokenType, String, Number, Endmarker);
create_node!(PythonNode, enum NodeType, file_input, foo);

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn it_works() {
        assert_eq!(2 + 2, 4);
        parse_python("bar");
        dbg!(TokenType::get_map());
        dbg!(NodeType::get_map());
        return
    }
}

struct PythonTokenizer {
}

impl PythonTokenizer {
    fn new_tok(&self) -> PythonToken {
        PythonToken {
            start: 0,
            length: 0,
            type_: TokenType::Endmarker,
        }
    }
}

impl parsa::Tokenizer<PythonToken> for PythonTokenizer {
    fn new(_code: &str) -> Self {
        Self {}
    }

    fn yield_next(&self) -> PythonToken {
        self.new_tok()
    }

}

create_parser!(parse_python, PythonTree, PythonNode,
               PythonToken, PythonTokenizer, TokenType, NodeType);
//create_parser!(parse_pythonx, PythonTreex, PythonNodex,
//               PythonToken, PythonTokenizer, TokenType, NodeType);
