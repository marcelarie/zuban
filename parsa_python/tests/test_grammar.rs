extern crate insta;

use parsa_python::*;

fn dedent(s: &'static str) -> String {
    let mut indent = usize::MAX;
    let lines: &Vec<_> = &s.split('\n').collect();
    for line in lines {
        if !line.trim().is_empty() {
            indent = std::cmp::min(indent, line.len() - line.trim_start().len());
        }
    }
    if indent == usize::MAX {
        return s.to_string();
    }
    let new_lines: Vec<_> = lines
        .iter()
        .map(|line| {
            if line.len() >= indent {
                &line[indent..]
            } else {
                line
            }
        })
        .collect();
    new_lines.join("\n")
}

fn tree_to_string(tree: PyTree) -> String {
    fn recurse(code: &mut String, node: &PyNode, depth: usize) {
        *code += &" ".repeat(depth);
        *code += &format!(
            "{}: {}-{}{}\n",
            node.type_str(),
            node.start(),
            node.end(),
            if node.is_leaf() {
                format!(" {:?}", node.as_code())
            } else {
                "".to_string()
            }
        );
        for c in node.iter_children() {
            assert_eq!(node.index, c.parent().unwrap().index);
            recurse(code, &c, depth + 1);
        }
    }

    let root = tree.root_node();
    assert!(root.parent().is_none());
    let mut code = String::new();
    recurse(&mut code, &root, 0);
    code
}

macro_rules! parametrize_snapshots {
    ($($name:ident : $input:expr;)*) => {$(
        #[test]
        fn $name() {
            let tree = parse($input.into());
            insta::assert_snapshot!(stringify!($name), tree_to_string(tree));
        }
    )*}
}

parametrize_snapshots!(
    empty: "".to_owned();
    very_simple: "1".to_owned();
    json: "{foo: 1}\n".to_owned();
    simple: dedent("
        a(bar:=1, foo=1);
        if a:
         1
        else: 2
        ");
    assignments_simple: dedent("
        a = 1
        a, b = 2
        a, b, = 2
        (a, b) = 2
        ((a, b)) = 2
        a = b = 3
        a = (b, c) = 3
        a = (b, c,) = 3
        a = b, c, = 3
        ");
    assignments_star: dedent("
        *foo = *bar = *baz
        i, *foo = ((j, [*bar, k])) = *baz
        *foo, *bar = yield baz
        ");
    assignments_aug: dedent("
        foo += 3
        foo.bar |= 3
        foo, bar += 3
        ");
    assignments_setitem: dedent("
        foo[0] += 3
        foo[0] = a = 3
        foo[a, b] = c
        foo[a::, b:f, c:=3] = c
        ");
    assignments_annotation: dedent("
        foo: bar
        foo: int = 3
        foo.bar: baz
        foo: bar = a = 4
        ");
    assignments_with_call: dedent("
        foo().bar = 1
        foo(1,).bar[1] = 1
        foo(a for a in b).bar[1] = 1
        ");
    for_stmt: dedent("
        for x in [1,2]:
            ...
        for *b in *c: ...
        for c.x, (*b, d.y[0]) in i, *j, k: ...
        ");
    comprehension_simple: dedent("
        [1 for a in b]
        [a for *a, c in b if c if d]
        foo(a for (*a, (*c, b),), in d)
        ");
    comprehension_fails: dedent("
        # fails
        [1 for a in *b]
        [1 for a in b if a else c]
        ");
    del_stmt: dedent("
        del foo
        del foo,
        del foo, bar.baz
        del foo[1]
        ");
    cls: dedent("
        class Foo(object):
            def __init__(self, /, f, *, g):
                pass

            @property
            def prop(self):
                pass
        ");
    import_names: dedent("
        import foo
        import foo.bar as baz
        import foo, bar, baz
        import foo.bar.baz
        ");
    import_name_fails: dedent("
        import foo,
        import foo, bar,
        ");
    import_from_normal: dedent("
        from foo import bar
        from foo import bar as baz
        from foo.bar.baz import *
        ");
    import_from_dotted: dedent("
        from . import (foo as baz, bar)
        from . import (foo, bar,)
        from ..foo import bar as baz
        from ... import *
        ");
    import_from_fails: dedent("
        from . import foo,
        from . import foo, bar,
        ");
    calls: dedent("
        foo(bar)
        foo(bar_foo=3)
        foo(bar_baz:=2)
        foo(**kwargs)
        ");
    nested_call: dedent("
        foo(1)()
        ");
    complex_calls: dedent("
        foo(bar, 3)
        foo(bar, *foo)
        foo(bar, *foo, name=3)
        foo(bar, *foo, name=3, **kwargs)
        foo(bar, *foo, name=3, **kwargs, b=3)
        foo(**kwargs, b=3, **foo)
        ");
    operators: dedent("
        f | d | c + 5 & 2 * 3 * 4 ** 5 ** 6
        ");
    simple_error_recovery: dedent("
        a + 3 +
        b = 3
        ");
    terminal_error_recovery: dedent("
        ?
        ");
    terminal_and_nonterminal_error_recovery: dedent("
        foo
        (1) + ? + c
        bar
        ");
    keyword_recovery: dedent("
        foo
        def
        bar
        else:
        baz
        ");
    match_simple: dedent("
        match foo:
            case ['a']: ...
            case _:
                pass
        ");
    soft_keyword_underscore: dedent("
        match _:
            case ['a']: _(_)
            case _:
                _
        match = 3
        ");
    match_underscore_with_as: dedent("
        match foo:
            case bar as _: pass
        ");
    match_class_pattern: dedent("
        match a:
            case A():
                pass
            case B(1, 2):
                pass
            case B(1, b=2):
                pass
            case B(a=1, b=2):
                pass
        ");
    dict_literal1: dedent("
        {1: 2}
        {**foo}
        {**foo, 1: 2}
        {**foo, **bar}
        ");
    function_simple: dedent("
        def foo(): ...
        def foo(a: 3): ...
        def foo(a, b,): ...
        def foo(a: f = 3, b=4): ...
        def foo(a=3, b=4): ...
        ");
    function_args: dedent("
        def foo(a, *, b=3): ...
        def foo(a=3, *, c): ...
        def foo(a, *, x, **kwargs): ...
        def foo(a, *args: int, **fooo: str,): ...
        def foo(a, /, b: i, c: i=3, **kwargs): ...
        def foo(*args): ...
        ");
    lambda_simple: dedent("
        lambda: ...
        lambda a: ...
        lambda a, b,: ...
        lambda a=3, b=4: ...
        ");
    lambda_slash: dedent("
        lambda a, /, b=3: ...
        lambda a=3, /,: ...
        lambda a=3, /: ...
        lambda a=3, /, *args, c=3: ...
        lambda a=3, /, c=3: ...
        lambda a, /, b, c=3, **kwargs: ...
        lambda a, a=3, /, c=3, *, d: ...
        ");
    lambda_args: dedent("
        lambda a, *, b=3: ...
        lambda a=3, *, c: ...
        lambda a, *, x, **kwargs: ...
        lambda a, *args, **fooo ,: ...
        lambda *args: ...
        ");
    lambda_kwargs: dedent("
        lambda **kwargs: ...
        lambda a=3, **kwargs: ...
        lambda a, **kwargs: ...
        lambda a, /, b, *, x, **kwargs: ...
        ");
    lambda_fails: dedent("
        lambda a, *, **kwargs: ...
        lambda a=3, b: ...
        lambda **a, b: ...
        lambda a=3, *,: ...
        lambda ,: ...
        ");
    lambda_fail_with_slash: dedent("
        lambda a=3, /, b: ...
        lambda a=3, /, b=3, c: ...
        lambda a, /, b=3, c: ...
        lambda z, a=3, d, /, b: ...
        ");
    with_items_simple: dedent("
        with foo as bar: ...
        with foo as bar, baz: ...
        ");
    with_items_parentheses: dedent("
        with (foo as bar): ...
        with (foo as bar, baz): ...
        with ((foo)): ...
        with (i for i in 42): ...
        with (yield y): ...
        ");
    string_literals_in_strings: dedent(r#"
        '"""'
        "'''"
        ""
        ''
        "#);
    bytes: dedent(r#"
        b'asdf\n\''
        br'\asdf'
        Br'asdf"""'
        bR''
        BR'\''
        "#);
    exception_group: dedent(r#"
        try:
            ...
        except* SpamError:
            ...
        except* FooError as e:
            ...
        except* (BarError, BazError) as e:
            ...
        "#);
    invalid_syntax: dedent(r#"
        def f():
            (x,) += 1
            z
        "#);
);
