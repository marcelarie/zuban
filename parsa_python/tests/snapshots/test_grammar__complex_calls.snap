---
source: parsa_python/tests/test_grammar.rs
expression: tree_to_string(tree)
---
file: 0-148
 stmt: 1-13
  simple_stmts: 1-13
   simple_stmt: 1-12
    expr_stmt: 1-12
     star_expressions: 1-12
      expression: 1-12
       bitwise_or: 1-12
        primary: 1-12
         Name: 1-4 "foo"
         Keyword: 4-5 "("
         arguments: 5-11
          named_expression: 5-8
           expression: 5-8
            bitwise_or: 5-8
             Name: 5-8 "bar"
          Keyword: 8-9 ","
          named_expression: 10-11
           expression: 10-11
            bitwise_or: 10-11
             Number: 10-11 "3"
         Keyword: 11-12 ")"
   Newline: 12-13 "\n"
 stmt: 13-28
  simple_stmts: 13-28
   simple_stmt: 13-27
    expr_stmt: 13-27
     star_expressions: 13-27
      expression: 13-27
       bitwise_or: 13-27
        primary: 13-27
         Name: 13-16 "foo"
         Keyword: 16-17 "("
         arguments: 17-26
          named_expression: 17-20
           expression: 17-20
            bitwise_or: 17-20
             Name: 17-20 "bar"
          Keyword: 20-21 ","
          starred_expression: 22-26
           Keyword: 22-23 "*"
           expression: 23-26
            bitwise_or: 23-26
             Name: 23-26 "foo"
         Keyword: 26-27 ")"
   Newline: 27-28 "\n"
 stmt: 28-51
  simple_stmts: 28-51
   simple_stmt: 28-50
    expr_stmt: 28-50
     star_expressions: 28-50
      expression: 28-50
       bitwise_or: 28-50
        primary: 28-50
         Name: 28-31 "foo"
         Keyword: 31-32 "("
         arguments: 32-49
          named_expression: 32-35
           expression: 32-35
            bitwise_or: 32-35
             Name: 32-35 "bar"
          Keyword: 35-36 ","
          starred_expression: 37-41
           Keyword: 37-38 "*"
           expression: 38-41
            bitwise_or: 38-41
             Name: 38-41 "foo"
          Keyword: 41-42 ","
          kwargs: 43-49
           kwarg_or_starred: 43-49
            Name: 43-47 "name"
            Keyword: 47-48 "="
            expression: 48-49
             bitwise_or: 48-49
              Number: 48-49 "3"
         Keyword: 49-50 ")"
   Newline: 50-51 "\n"
 stmt: 51-84
  simple_stmts: 51-84
   simple_stmt: 51-83
    expr_stmt: 51-83
     star_expressions: 51-83
      expression: 51-83
       bitwise_or: 51-83
        primary: 51-83
         Name: 51-54 "foo"
         Keyword: 54-55 "("
         arguments: 55-82
          named_expression: 55-58
           expression: 55-58
            bitwise_or: 55-58
             Name: 55-58 "bar"
          Keyword: 58-59 ","
          starred_expression: 60-64
           Keyword: 60-61 "*"
           expression: 61-64
            bitwise_or: 61-64
             Name: 61-64 "foo"
          Keyword: 64-65 ","
          kwargs: 66-82
           kwarg_or_starred: 66-72
            Name: 66-70 "name"
            Keyword: 70-71 "="
            expression: 71-72
             bitwise_or: 71-72
              Number: 71-72 "3"
           Keyword: 72-73 ","
           double_starred_expression: 74-82
            Keyword: 74-76 "**"
            expression: 76-82
             bitwise_or: 76-82
              Name: 76-82 "kwargs"
         Keyword: 82-83 ")"
   Newline: 83-84 "\n"
 stmt: 84-122
  simple_stmts: 84-122
   simple_stmt: 84-121
    expr_stmt: 84-121
     star_expressions: 84-121
      expression: 84-121
       bitwise_or: 84-121
        primary: 84-121
         Name: 84-87 "foo"
         Keyword: 87-88 "("
         arguments: 88-120
          named_expression: 88-91
           expression: 88-91
            bitwise_or: 88-91
             Name: 88-91 "bar"
          Keyword: 91-92 ","
          starred_expression: 93-97
           Keyword: 93-94 "*"
           expression: 94-97
            bitwise_or: 94-97
             Name: 94-97 "foo"
          Keyword: 97-98 ","
          kwargs: 99-120
           kwarg_or_starred: 99-105
            Name: 99-103 "name"
            Keyword: 103-104 "="
            expression: 104-105
             bitwise_or: 104-105
              Number: 104-105 "3"
           Keyword: 105-106 ","
           double_starred_expression: 107-115
            Keyword: 107-109 "**"
            expression: 109-115
             bitwise_or: 109-115
              Name: 109-115 "kwargs"
           Keyword: 115-116 ","
           kwarg: 117-120
            Name: 117-118 "b"
            Keyword: 118-119 "="
            expression: 119-120
             bitwise_or: 119-120
              Number: 119-120 "3"
         Keyword: 120-121 ")"
   Newline: 121-122 "\n"
 stmt: 122-148
  simple_stmts: 122-148
   simple_stmt: 122-147
    expr_stmt: 122-147
     star_expressions: 122-147
      expression: 122-147
       bitwise_or: 122-147
        primary: 122-147
         Name: 122-125 "foo"
         Keyword: 125-126 "("
         arguments: 126-146
          kwargs: 126-146
           double_starred_expression: 126-134
            Keyword: 126-128 "**"
            expression: 128-134
             bitwise_or: 128-134
              Name: 128-134 "kwargs"
           Keyword: 134-135 ","
           kwarg: 136-139
            Name: 136-137 "b"
            Keyword: 137-138 "="
            expression: 138-139
             bitwise_or: 138-139
              Number: 138-139 "3"
           Keyword: 139-140 ","
           double_starred_expression: 141-146
            Keyword: 141-143 "**"
            expression: 143-146
             bitwise_or: 143-146
              Name: 143-146 "foo"
         Keyword: 146-147 ")"
   Newline: 147-148 "\n"
 Endmarker: 148-148 ""

